### PR TITLE
Refactor architecture

### PR DESCRIPTION
**Pull request by aperezg**
Adding all the improvement that we're seeing on our companies and personal projects that working better to use the hexagonal architecture on Go

### CODE REVIEW

File: contexts-architecture/users/cmd/users-api/internal/server/http/handlers.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ func getUserHandlerBuilder(
 		if authorizedUser.ID != userID {
 			errorMsg := fmt.Sprintf("the user %s cannot read the data of user %s", authorizedUser.ID, userID)
 			c.JSON(http.StatusForbidden, gin.H{"error": errorMsg})
+			return
```
Comment (joanlopez) [inline contexts-architecture/users/cmd/users-api/internal/server/http/handlers.go:26]: Good catch!! 👍

File: hexagonal-architecture/cmd/http/main.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package main
+
+import (
+	"log"
+
+	"github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal/platform/bootstrap"
```
Comment (joanlopez) [inline hexagonal-architecture/cmd/http/main.go]: Can you move all the code under `/internal/platform/bootstrap` into `cmd/http/bootstrap`, please? 🙏

File: hexagonal-architecture/internal/counter.go [hunk 3]
```diff
@@ -12,6 +13,11 @@ const (
 	defaultCounterValue = 0
 )
 
+var (
+	ErrCounterNotFound = errors.New("counter not found")
+	ErrNameTooShort    = errors.New("the receive name is too short")
```
Comment (joanlopez) [inline hexagonal-architecture/internal/counter.go]:
```suggestion
	ErrNameTooShort    = errors.New("the received name is too short")
```

or even better

```suggestion
	ErrNameTooShort    = errors.New("counter name is too short")
```

File: hexagonal-architecture/internal/counter.go [hunk 4]
```diff
@@ -20,12 +26,12 @@ type Counter struct {
 	BelongsTo  string
 }
 
-func NewCounter(name, belongsTo string) (*Counter, error) {
+func NewCounter(name, belongsTo string) (Counter, error) {
 	if len(name) < minNameLength {
-		return nil, errors.NewWrongInput("counter name %s is too short", name)
+		return Counter{}, fmt.Errorf("min value is: %d: %w", minNameLength, ErrNameTooShort)
```
Comment (joanlopez) [inline hexagonal-architecture/internal/counter.go:31]: I find the new approach much more beautiful 👍

File: hexagonal-architecture/internal/creating/service.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package creating
+
+import (
+	counters "github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal"
+)
+
+type DefaultService interface {
```
Comment (joanlopez) [inline hexagonal-architecture/internal/creating/service.go]: If I'm not wrong, the discussed proposal was the other way round:

- `DefaultService` for the default implementation (the `struct`).
- `Service` for the service as such (the behavior definition, so the `interface`).

Can you confirm this, please? 🤔  Thanks!!!

File: hexagonal-architecture/internal/fetching/service.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package fetching
+
+import (
+	counters "github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal"
+)
+
+type DefaultService interface {
```
Comment (joanlopez) [inline hexagonal-architecture/internal/fetching/service.go]: Same as previously mentioned.

File: hexagonal-architecture/internal/incrementing/service.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package incrementing
+
+import (
+	counters "github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal"
+)
+
+type DefaultService interface {
```
Comment (joanlopez) [inline hexagonal-architecture/internal/incrementing/service.go]: Same as previously mentioned.

File: hexagonal-architecture/internal/platform/server/http/handler/users/handler.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+	"github.com/friendsofgo/errors"
+	counters "github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal"
+	"github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal/creating"
+	"github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal/fetching"
+	server "github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal/platform/server/http"
+	"github.com/gin-gonic/gin"
+)
+
+func GetUserHandlerBuilder(
```
Comment (joanlopez) [inline hexagonal-architecture/internal/platform/server/http/handler/users/handler.go]: Nitpicky (so you can ignore) but just to mention it:

What do you think about removing the `Builder` suffix from the HTTP handler builders? 
The intention would just to be coherent with the naming conventions used by the Gin.

Note that [it don't use any suffix for built-in middlewares](https://github.com/gin-gonic/gin/blob/master/logger.go#L183) which at the end also return a `HandlerFunc`.

File: hexagonal-architecture/internal/user.go [hunk 9]
```diff
@@ -1,22 +1,27 @@
...
+
 	"golang.org/x/crypto/bcrypt"
 
-	"github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/internal/errors"
 	"github.com/friendsofgo/go-architecture-examples/hexagonal-architecture/kit/ulid"
 )
 
+var (
+	ErrUserNotFound = errors.New("user not found")
+	ErrCreatingUser = errors.New("err creating user")
```
Comment (joanlopez) [inline hexagonal-architecture/internal/user.go]:
```suggestion
	ErrCreatingUser = errors.New("error happened while creating a new user")
```

**Review by joanlopez — COMMENTED** [review]
Awesome work!!! 🙏 🙏 🙏 

I just left few comments, note that some of those comments & suggestions are a bit nitpicky so you can directly ignore them and merge it.

File: hexagonal-architecture/internal/creating/service.go (hunk 5 above)
Comment (aperezg) [reply hexagonal-architecture/internal/creating/service.go]: Yes you're totally right!

**Review by joanlopez — APPROVED** [review]
LGTM!!! Thanks for your time considering my comments ❤️

**Pull request merged**